File: src/policy/language-gadget-guard.ts
/**
 * Language-specific type confusion / deserialization gadget detection.
 */
import { walkStringLeaves } from './arg-leaf-walker.js';
import { deobfuscateRecursive } from '../utils/payload-normalizer.js';
import type { CallContext, PolicyDecision } from './policy-types.js';

export const LANGUAGE_GADGET_PATTERNS: RegExp[] = [
  /\b(?:pickle\.loads?|cPickle\.loads?|dill\.loads?)\b/i,
  /\b__reduce__\b|\bPYCC\b/i,
  /\bObjectInputStream\b|\breadObject\s*\(\s*\)/i,
  /\bjava\.io\.(?:Serializable|ObjectInputStream)\b/i,
  /\b(?:org\.apache\.commons\.collections|ysoserial)/i,
  /\b(?:node-serialize|node-serializer)\b/i,
  /\bunserialize\s*\(\s*['"]/i,
  /\b__wakeup\b|\b__destruct\b.*\beval\b/i,
  /\bMarshal\.load\b|\byaml\.unsafe_load\b|\byaml\.load\s*\([^)]*Loader\s*=\s*yaml\.UnsafeLoader/i,
  /!!python\/object/i,
  /\bcos(?:\\n|\s*\n)\s*system(?:\\n|\s*\n)/i,
  /\bcos\s+system\s*\(/i,
  /<!ENTITY\b|<!DOCTYPE[^>]{0,200}\bENTITY\b/i,
  /\bSYSTEM\s+["']file:\/\//i,
  /\bphp:\/\/filter\b/i,
  /\bRuntime\.getRuntime\s*\(\s*\)\.exec\b/i,
  /\bProcessBuilder\s*\(/i,
  /\bScriptEngineManager\b.*\bgetEngineByName\b/i,
  /\bBinaryFormatter\b.*\bDeserialize\b/i,
  /\bViewState\b.*\bMAC\b/i,
  /\br00t\.me\b/i,
  /application\/x-(?:java-serialized-object|python-serialized)/i,
];

export function evaluateLanguageGadgetGuard(ctx: CallContext): PolicyDecision | null {
  const blob = walkStringLeaves(ctx.arguments ?? {})
    .map((l) => deobfuscateRecursive(l.value))
    .join('\n');

  if (!blob.trim()) return null;

  for (const pattern of LANGUAGE_GADGET_PATTERNS) {
    if (pattern.test(blob)) {
      return {
        action: 'block',
        rule: 'semantic-language-gadget',
        reason: 'Language-specific deserialization or gadget chain pattern in arguments',
      };
    }
  }

  return null;
}

File: adversarial-harness/python/policy_engine/language_gadget_guard.py
"""Language-specific deserialization gadget detection."""

from __future__ import annotations

import json
import re
from typing import Optional

from .types import CallContext, PolicyDecision

GADGET_PATTERNS = [
    re.compile(r"\b(?:pickle\.loads?|cPickle|__reduce__)\b", re.I),
    re.compile(r"\bObjectInputStream\b|\breadObject\s*\(\s*\)", re.I),
    re.compile(r"\b(?:ysoserial|node-serialize)\b", re.I),
    re.compile(r"\bunserialize\s*\(", re.I),
    re.compile(r"\bMarshal\.load\b|\byaml\.unsafe_load\b", re.I),
]


def evaluate_language_gadget_guard(ctx: CallContext) -> Optional[PolicyDecision]:
    blob = json.dumps(ctx.arguments or {}, ensure_ascii=False)
    if any(p.search(blob) for p in GADGET_PATTERNS):
        return PolicyDecision(
            "block",
            "semantic-language-gadget",
            "Language-specific gadget pattern in arguments",
        )
    return None
